x-pack/plugins/reporting/server/export_types/common/pdf/pdfmaker.ts

```diff
@@ -5,52 +5,70 @@
-import concat from 'concat-stream';
+import { SerializableRecord } from '@kbn/utility-types';
 import _ from 'lodash';
 import path from 'path';
-import Printer from 'pdfmake';
+import { Worker } from 'worker_threads';
 import { Content, ContentImage, ContentText } from 'pdfmake/interfaces';
 import type { Layout } from '../../../../../screenshotting/server';
-import { getDocOptions, REPORTING_TABLE_LAYOUT } from './get_doc_options';
+import { REPORTING_TABLE_LAYOUT } from './get_doc_options';
 import { getFont } from './get_font';
-import { getTemplate } from './get_template';
-
-const assetPath = path.resolve(__dirname, '..', '..', 'common', 'assets');
-const tableBorderWidth = 1;
+import {
+  headingHeight,
+  pageMarginBottom,
+  pageMarginTop,
+  pageMarginWidth,
+  subheadingHeight,
+  tableBorderWidth,
+} from './constants';
+import { PdfWorkerOutOfMemoryError } from './pdf_generate_errors';
+
+import type { PdfWorkerData } from './worker';
 
 export class PdfMaker {
-  private _layout: Layout;
+  _layout: Layout;
   private _logo: string | undefined;
   private _title: string;
   private _content: Content[];
-  private _printer: Printer;
-  private _pdfDoc: PDFKit.PDFDocument | undefined;
 
-  constructor(layout: Layout, logo: string | undefined) {
-    const fontPath = (filename: string) => path.resolve(assetPath, 'fonts', filename);
-    const fonts = {
-      Roboto: {
-        normal: fontPath('roboto/Roboto-Regular.ttf'),
-        bold: fontPath('roboto/Roboto-Medium.ttf'),
-        italics: fontPath('roboto/Roboto-Italic.ttf'),
-        bolditalics: fontPath('roboto/Roboto-Italic.ttf'),
-      },
-      'noto-cjk': {
-        // Roboto does not support CJK characters, so we'll fall back on this font if we detect them.
-        normal: fontPath('noto/NotoSansCJKtc-Regular.ttf'),
-        bold: fontPath('noto/NotoSansCJKtc-Medium.ttf'),
-        italics: fontPath('noto/NotoSansCJKtc-Regular.ttf'),
-        bolditalics: fontPath('noto/NotoSansCJKtc-Medium.ttf'),
-      },
-    };
+  private worker?: Worker;
+
+  protected workerModulePath = path.resolve(__dirname, './worker.js');
+
+  /**
+   * The maximum heap size for old memory region of the worker thread.
+   *
+   * @note We need to choose a sane number given that we need to load a Kibana
+   * node environment, all the library code and the subsequent buffers that
+   * could result from generating a PDF.
+   *
+   * Local testing indicates that to trigger an OOM event for the worker we need
+   * to exhaust not only heap but also any compression optimization and fallback
+   * swap space.
+   *
+   * With this value we are able to generate PDFs in excess of 5000x5000 pixels
+   * at which point issues other than memory start to show like glitches in the
+   * image.
+   *
+   * TODO: Should we consider making this number dynamic? It is difficult to
+   * know how we'd describe this to users in the most helpful way.
+   */
+  protected workerMaxOldHeapSizeMb = 128;
```

Comment:
This value feels similar in purpose to the `--max-old-space-size` which is configured in the `config/node.options` file. If we treat it that way, then it should be scalable based on the amount of system RAM.

Reply:
Gotcha. Perhaps what we need is a heuristic that says: the worker can be allocated: `Math.min(currentOldHeap, 512)`.

What do you think?

Reply:
That, or a formula similar to it, sounds reasonable. My understanding is that we should not let Node use its default value, because that value is not connected to the `--max-old-space-size` setting

Reply:
After more study on worker threads, I am in favor of leaving it the way you have it.

> With this value we are able to generate PDFs in excess of 5000x5000 pixels

This is plenty. Knowing this value gives us something tangible to document (in a different PR).

Making it adjustable or higher adds a ton of risk that the worker thread could get enough memory to crash the main thread.